server/core/src/logical_locking.cpp

```diff
@@ -0,0 +1,305 @@
+            irods::log(LOG_WARNING, fmt::format(
+                "[{}:{}] - failed to parse data_status value [{}]",
+                __FUNCTION__, __LINE__, data_status_str));
+
+            throw;
+        }
+    } // get_original_replica_status_impl
+
+    auto insert_data_status(const std::uint64_t _data_id) -> void
+    {
+        auto entry = rst::at(_data_id);
+
+        const auto logical_path = rst::get_logical_path(_data_id);
+
+        for (auto& json_replica : entry) {
+            const auto replica_number = std::stoi(json_replica.at("after").at("data_repl_num").get<std::string>());
+
+            irods::log(LOG_DEBUG, fmt::format(
+                "[{}:{}] - replica:[{}]",
+                __FUNCTION__, __LINE__, json_replica.dump()));
+
+            if (const auto original_status = get_original_replica_status_impl(json_replica); -1 != original_status) {
+                irods::log(LOG_NOTICE, fmt::format(
+                    "[{}:{}] - existing data_status entry; "
+                    "[data_id:[{}], repl_num:[{}], status:[{}]]",
+                    __FUNCTION__, __LINE__, _data_id, replica_number, original_status));
+
+                continue;
+            }
+
+            const auto old_status = rst::get_property(_data_id, replica_number, "data_is_dirty");
+
+            const auto original_status = json{{"original_status", old_status}};
+            const auto update = json{{"data_status", original_status.dump()}};
+
+            irods::log(LOG_DEBUG, fmt::format(
+                "[{}:{}] - update:[{}]",
+                __FUNCTION__, __LINE__, update.dump()));
+
+            rst::update(_data_id, replica_number, update);
+        }
+    } // insert_data_status
+
+    auto remove_data_status(const std::uint64_t _data_id) -> void
+    {
+        for (const auto& json_replica : rst::at(_data_id)) {
+            const auto replica_number = std::stoi(json_replica.at("after").at("data_repl_num").get<std::string>());
+            rst::update(_data_id, replica_number, json{{"data_status", ""}});
+        }
+    } // remove_data_status
+
+    auto restore_replica_statuses(
+        const std::uint64_t _data_id,
+        const int           _replica_number) -> void
+    {
+        auto entry = rst::at(_data_id);
+
+        const auto logical_path = rst::get_logical_path(_data_id);
+
+        for (auto& json_replica : entry) {
+            const auto replica_number = std::stoi(json_replica.at("after").at("data_repl_num").get<std::string>());
+
+            if (_replica_number == replica_number) {
+                continue;
+            }
+
+            const auto original_status = get_original_replica_status_impl(json_replica);
+
+            if (-1 == original_status) {
+                irods::log(LOG_ERROR, fmt::format(
+                    "[{}:{}] - failed to restore status for replica; "
+                    "original status not found in data_status column "
+                    "[data_id=[{}], repl num=[{}]]",
```

Comment:
repl_num  with underscore?     fine either way, just want it consistent.

Reply:
Done